### PR TITLE
Clarify documentation related to `Array` and `ArrayLayout`

### PR DESCRIPTION
**Pull request by whitequark**

### CODE REVIEW

File: docs/guide.rst [hunk 1]
```diff
@@ -811,7 +811,11 @@ Crucially, this means that any Python object can be added to an array; the only
 
 .. warning::
 
-    Each time an array proxy object with :py:`n` elements is used in an expression, it generates a multiplexer with :py:`n` branches. However, using :py:`k` of such array proxy objects in an expression generates a multiplexer with :py:`n**k` branches. This can generate extremely large circuits that may quickly exhaust the resources of the synthesis target or even the available RAM.
```
Comment (wanda-phi) [inline docs/guide.rst:814]: I don't think that's true anymore

Comment (whitequark) [reply docs/guide.rst:814]: Can you elaborate?

Comment (wanda-phi) [reply docs/guide.rst:814]: it got fixed with the new NIR-based RTLIL backend

Comment (whitequark) [reply docs/guide.rst:814]: Oh! Right

**Pull request merged**